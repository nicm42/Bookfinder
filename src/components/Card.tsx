import PropTypes, { string } from 'prop-types';
import * as Styled from './Card.styles';
import cover from '../images/generic-book-cover.jpg';

interface CardProps {
  card: any;
}

const Card = ({ card }: CardProps) => {
  //The covers automatically have a curled edge, which I don't like
  //so we'll remove it
  let bookCover;
  if (card.volumeInfo.imageLinks) {
    bookCover = card.volumeInfo.imageLinks.thumbnail.replace('&edge=curl', '');
  }

  //If there's more than one author, we need to format them
  //otherwise it puts the names right next to each other with no comma or space
  const { authors } = card.volumeInfo;
  let formattedAuthors = '';
  if (authors) {
    formattedAuthors = authors.map((author: string[], index: number) => {
      if (index > 0) {
        if (index === authors.length - 1) {
          return ` & ${author}`;
        }
        return `, ${author}`;
      }
      return `by ${author}`;
    });
  }

  //Making ternary operators variables so the return is easier to read
  const coverSrc = card.volumeInfo.imageLinks ? bookCover : cover;
  const coverAlt = card.volumeInfo.title
    ? `Cover for ${card.volumeInfo.title}`
    : 'Book cover';
  const titleText = card.volumeInfo.title
    ? card.volumeInfo.title
    : 'Title missing';
  const authorText = card.volumeInfo.authors
    ? formattedAuthors
    : 'Author missing';
  const publisherText = card.volumeInfo.publisher
    ? `Publisher: ${card.volumeInfo.publisher}`
    : 'Publisher missing';
  const linkText = card.volumeInfo.title ? card.volumeInfo.title : 'this book';
  const link = card.volumeInfo.infoLink ? (
    <a
      href={card.volumeInfo.infoLink}
      title={`More information about ${linkText}`}
    >
      More information
    </a>
  ) : (
    <p>Link missing</p>
  );

  return (
    <Styled.CardDiv data-testid="cardDiv">
      <Styled.Cover src={coverSrc} alt={coverAlt} />
      <Styled.Info>
        <Styled.Title data-testid="title">{titleText}</Styled.Title>
        <Styled.Author data-testid="author">{authorText}</Styled.Author>
        <Styled.Publisher data-testid="publisher">
          {publisherText}
        </Styled.Publisher>
        {link}
      </Styled.Info>
    </Styled.CardDiv>
  );
};

Card.propTypes = {
  card: PropTypes.shape({
    volumeInfo: PropTypes.shape({
      imageLinks: PropTypes.shape({
        thumbnail: PropTypes.string,
      }),
      title: PropTypes.string,
      authors: PropTypes.arrayOf(string),
      publisher: PropTypes.string,
      infoLink: PropTypes.string,
    }),
  }).isRequired,
};

export default Card;
