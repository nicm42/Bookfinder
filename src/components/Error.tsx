import PropTypes from 'prop-types';
import * as Styled from './Error.styles';

interface ErrorProps {
  errorMessage: string;
}

const Error = ({ errorMessage }: ErrorProps) => (
  <Styled.Error data-testid="error">{errorMessage}</Styled.Error>
);

Error.propTypes = {
  errorMessage: PropTypes.string.isRequired,
};

export default Error;
