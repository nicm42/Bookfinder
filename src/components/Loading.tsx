import * as Styled from './Loading.styles';

const Loading = () => <Styled.Loading data-testid="loading" />;

export default Loading;
