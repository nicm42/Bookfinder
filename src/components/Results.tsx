import PropTypes from 'prop-types';
import * as Styled from './Results.styles';

interface ResultsProps {
  resultStart: number;
  resultEnd: number;
}

const Results = ({ resultStart, resultEnd }: ResultsProps) => (
  <Styled.ResultsCount>
    Showing books {resultStart}-{resultEnd}
  </Styled.ResultsCount>
);

Results.propTypes = {
  resultStart: PropTypes.number.isRequired,
  resultEnd: PropTypes.number.isRequired,
};

export default Results;
