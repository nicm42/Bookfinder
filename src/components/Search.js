import React, { useState } from 'react';
import { Label, Input, Submit } from './Search.styles';

const Search = () => {
  const [searchText, setSearchText] = useState();

  const runSearch = () => {
    console.log(searchText);
    //TODO send this to Google Books API
  };

  return (
    <>
      <Label htmlFor="search-input">Search for a book </Label>
      <Input
        type="search"
        id="search-input"
        name="search"
        required
        autocomplete="on"
        onChange={(e) => setSearchText(e.target.value)}
      />
      <Submit onClick={runSearch}>Search</Submit>
    </>
  );
};

export default Search;
