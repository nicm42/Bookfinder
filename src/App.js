import { useEffect } from 'react';
import Search from './components/Search';

const App = () => {
  useEffect(() => {
    document.title = 'Bookfinder';
  }, []);

  return <Search />;
};

export default App;
