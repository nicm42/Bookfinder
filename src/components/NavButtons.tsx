import * as Styled from '../App.style';

interface NavButtonProps {
  isPreviousResults: boolean;
  isMoreResults: boolean;
  goBack: () => void;
  searchAgain: () => void;
}

const NavButtons = ({
  isPreviousResults,
  isMoreResults,
  goBack,
  searchAgain,
}: NavButtonProps) => (
  <Styled.PrevNext>
    <Styled.Previous
      disabled={!isPreviousResults}
      onClick={goBack}
      isPreviousResults={isPreviousResults}
    >
      Previous
    </Styled.Previous>

    <Styled.Next
      disabled={!isMoreResults}
      onClick={searchAgain}
      isMoreResults={isMoreResults}
    >
      Next
    </Styled.Next>
  </Styled.PrevNext>
);

export default NavButtons;
