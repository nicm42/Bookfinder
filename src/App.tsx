import { useState, useEffect, useRef } from 'react';
import axios from 'axios';
import Search from './components/Search';
import Loading from './components/Loading';
import Error from './components/Error';
import Results from './components/Results';
import NavButtons from './components/NavButtons';
import Card from './components/Card';
//import DataContext from './contexts/DataContext';
import ButtonContext from './contexts/ButtonContext';
import * as Styled from './App.styles';
//import { testCards } from './dummyCardData'; //uncomment to load cards without using API

const App = () => {
  const [cardData, setCardData] = useState<any[]>([]);
  //const [cardData, setCardData] = useState<any[]>(testCards); //uncomment to load cards without using API
  const [results, setResults] = useState<any[]>([]);
  const [isLoading, setIsLoading] = useState<Boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>('');

  const [searchText, setSearchText] = useState<string>('');
  const [searchType, setSearchType] = useState<string>('');

  const [showButtons, setShowButtons] = useState<boolean>(false);
  const [isPreviousResults, setIsPreviousResults] = useState<boolean>(false);
  const [isMoreResults, setIsMoreResults] = useState<boolean>(false);
  const [pageNumber, setPageNumber] = useState<number>(0);

  const [resultStart, setResultStart] = useState<number>(-9);
  const [resultEnd, setResultEnd] = useState<number>(0);
  const [totalItems, setTotalItems] = useState<number>(0);

  const didMountRef = useRef<boolean>(false);

  /* Google Books API automatically returns 10 books 
    but since I'm using that 10 all over the place we need a constant
    - and to see maxResults in case they change that default*/
  const resultsPerPage: number = 10;
  let startIndex: number;

  useEffect(() => {
    if (didMountRef.current) {
      if (pageNumber === 1) {
        setIsPreviousResults(false);
      } else {
        setIsPreviousResults(true);
      }
    } else {
      didMountRef.current = true;
    }
  }, [pageNumber]);

  const getData = async (search: string, type: string, start: number) => {
    //Need to re-set everything if this is a new search
    if (start === 0) {
      setResultStart(1 - resultsPerPage);
      setResultEnd(0);
      setPageNumber(1);
      setIsMoreResults(false);
      setIsPreviousResults(false);
      setTotalItems(0);
      setResults([]);
      //set these two, so we can use them for another search
      setSearchText(search);
      setSearchType(type);
    } else {
      setIsPreviousResults(true);
    }

    setShowButtons(false);
    setIsLoading(true);
    //clear these, ready for a new search
    setCardData([]);
    setErrorMessage('');

    try {
      const api = 'https://www.googleapis.com/books/v1/volumes?q=';
      const response = await axios.get(
        `${api}${type}:%22${search}%22&startIndex=${start}&maxResults=10`
      );

      //Uncomment line below to test API errors
      //const response = await axios.get(`http://httpstat.us/404`);
      //Uncomment one of the lines below to test API timeout
      /* const response = await axios.get(`http://httpstat.us/504`, {
        timeout: 2,*/
      /* const response = await axios.get(`http://httpstat.us/408`); */

      setTotalItems(response.data.totalItems);

      if (!response.data.items) {
        if (type === 'intitle') {
          setErrorMessage(`No books were found with the title ${search} :(`);
        }
        if (type === 'inauthor') {
          setErrorMessage(`No books were found with the author ${search} :(`);
        }
      } else {
        //There is data!
        setCardData(response.data.items);
        //setResultCount(response.data.totalItems);
        //console.log(response.data.totalItems);
        setResultStart((previousValue) => previousValue + resultsPerPage);
        //Save these results so we can use them later if we need to go back to them
        setResults((arr) => [...arr, response.data.items]);
        if (response.data.totalItems > resultsPerPage) {
          setShowButtons(true);
          //Save these results so we can use them later if we need to go back to them
          if (start + 1 + resultsPerPage < response.data.totalItems) {
            setIsMoreResults(true);
          } else {
            setIsMoreResults(false);
          }
        } else {
          setIsMoreResults(false);
        }
        setResultEnd(
          (previousValue) => previousValue + response.data.items.length
        );
      }
    } catch (error) {
      console.log(error);
      if (
        error.message.replace(/ .*/, '') === 'timeout' ||
        error.request.status === 408 ||
        error.request.status === 504
      ) {
        setErrorMessage('The request timed out. Please try again later');
      } else {
        setErrorMessage(
          `Something went wrong :( Please speak to the developer with the search term: '${search}' and the error message: '${error.response.status} ${error.response.statusText}'`
        );
      }
    } finally {
      setIsLoading(false);
    }
  };

  const searchAgain = () => {
    //If we already have this data, then just show that
    startIndex = resultStart + resultsPerPage - 2;
    if (results.length >= pageNumber + 1) {
      window.scrollTo({ top: 0 }); //scroll back up, otherwise it's not clear anything has changed
      setCardData(results[pageNumber]);
      if (resultEnd + resultsPerPage > totalItems) {
        setIsMoreResults(false);
      }
      setResultStart((previousValue) => previousValue + resultsPerPage);
      setResultEnd(
        (previousValue) => previousValue + results[pageNumber].length
      );
    }
    setPageNumber((previousValue) => previousValue + 1);
    //Otherwise get it from the API
    if (results.length < pageNumber + 1) {
      getData(searchText, searchType, startIndex);
    }
  };

  const goBack = () => {
    window.scrollTo({ top: 0 }); //scroll back up, otherwise it's not clear anything has changed
    startIndex = resultStart - resultsPerPage - 2;
    setCardData(results[pageNumber - 2]);
    setIsMoreResults(true);
    setPageNumber((previousValue) => previousValue - 1);
    setResultStart((previousValue) => previousValue - resultsPerPage);
    /* setResultEnd(
      (previousValue) => previousValue - results[pageNumber - 2].length
    ); */
    //Take the last set of results and round it down to the nearest 10
    //But if it's 10, 20, 30 etc then just need to take 10 off it
    if (resultEnd % 10 === 0) {
      setResultEnd((previousValue) => previousValue - resultsPerPage);
    } else {
      setResultEnd(
        (previousValue) =>
          Math.floor(previousValue / resultsPerPage) * resultsPerPage
      );
    }
  };

  return (
    <>
      <header>
        <Styled.Header>Book Search</Styled.Header>
      </header>

      <main>
        <Search getData={getData} />
        {errorMessage && <Error errorMessage={errorMessage} />}
        {resultEnd > 0 && (
          <Results resultStart={resultStart} resultEnd={resultEnd} />
        )}
        {showButtons && (
          <ButtonContext.Provider value={{ isPreviousResults, isMoreResults }}>
            <NavButtons goBack={goBack} searchAgain={searchAgain} />
          </ButtonContext.Provider>
        )}
        {isLoading && <Loading />}
        <Styled.Books>
          {cardData &&
            cardData.map((card) => (
              <Card card={card} key={card.id} data-testid="card" />
            ))}
        </Styled.Books>
        {showButtons && (
          <ButtonContext.Provider value={{ isPreviousResults, isMoreResults }}>
            <NavButtons goBack={goBack} searchAgain={searchAgain} />
          </ButtonContext.Provider>
        )}
      </main>
    </>
  );
};

export default App;
