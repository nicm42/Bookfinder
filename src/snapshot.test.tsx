import renderer from 'react-test-renderer';
import App from './App';
import Search from './components/Search';
import Card from './components/Card';
import Loading from './components/Loading';
import Error from './components/Error';
import Results from './components/Results';
import NavButtons from './components/NavButtons';
import cards from './dummyCardData';

describe('Snapshot tests', () => {
  const getData = jest.fn();
  const goBack = jest.fn();
  const searchAgain = jest.fn();

  it('renders the search correctly', () => {
    const tree = renderer.create(<Search getData={getData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the cards correctly', () => {
    const tree = renderer.create(<Card card={cards.items[0]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders loading spinner correctly', () => {
    const tree = renderer.create(<Loading />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders error message correctly', () => {
    const tree = renderer
      .create(<Error errorMessage="This is an error message" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders results text correctly', () => {
    const tree = renderer
      .create(<Results resultStart={1} resultEnd={10} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nav buttons correctly for both buttons', () => {
    const tree = renderer
      .create(
        <NavButtons
          isPreviousResults
          isMoreResults
          searchAgain={searchAgain}
          goBack={goBack}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nav buttons correctly for just next button', () => {
    const tree = renderer
      .create(
        <NavButtons
          isPreviousResults={false}
          isMoreResults
          searchAgain={searchAgain}
          goBack={goBack}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nav buttons correctly for just previous button', () => {
    const tree = renderer
      .create(
        <NavButtons
          isPreviousResults
          isMoreResults={false}
          searchAgain={searchAgain}
          goBack={goBack}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nav buttons correctly for no buttons', () => {
    const tree = renderer
      .create(
        <NavButtons
          isPreviousResults={false}
          isMoreResults={false}
          searchAgain={searchAgain}
          goBack={goBack}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the app correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
