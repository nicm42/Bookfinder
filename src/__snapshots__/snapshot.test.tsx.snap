// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests renders error message correctly 1`] = `
<p
  className="Errorstyles__Error-b0ye4i-0 dipzyz"
  data-testid="error"
>
  This is an error message
</p>
`;

exports[`Snapshot tests renders loading spinner correctly 1`] = `
<div
  className="Loadingstyles__LoadingDiv-yove8p-0 dPOaed"
>
  <div
    className="Loadingstyles__Loading-yove8p-1 AJFPM"
    data-testid="loading"
  />
</div>
`;

exports[`Snapshot tests renders nav buttons correctly for both buttons 1`] = `
<div
  className="NavButtonsstyles__PrevNext-tjoruy-0 fPtyJB"
>
  <button
    className="NavButtonsstyles__Previous-tjoruy-1 iyEMDy"
    disabled={false}
    onClick={[MockFunction]}
  >
    Previous
  </button>
  <button
    className="NavButtonsstyles__Next-tjoruy-2 frJrFi"
    disabled={false}
    onClick={[MockFunction]}
  >
    Next
  </button>
</div>
`;

exports[`Snapshot tests renders nav buttons correctly for just next button 1`] = `
<div
  className="NavButtonsstyles__PrevNext-tjoruy-0 fPtyJB"
>
  <button
    className="NavButtonsstyles__Previous-tjoruy-1 mipfr"
    disabled={true}
    onClick={[MockFunction]}
  >
    Previous
  </button>
  <button
    className="NavButtonsstyles__Next-tjoruy-2 frJrFi"
    disabled={false}
    onClick={[MockFunction]}
  >
    Next
  </button>
</div>
`;

exports[`Snapshot tests renders nav buttons correctly for just previous button 1`] = `
<div
  className="NavButtonsstyles__PrevNext-tjoruy-0 fPtyJB"
>
  <button
    className="NavButtonsstyles__Previous-tjoruy-1 iyEMDy"
    disabled={false}
    onClick={[MockFunction]}
  >
    Previous
  </button>
  <button
    className="NavButtonsstyles__Next-tjoruy-2 eYzWCn"
    disabled={true}
    onClick={[MockFunction]}
  >
    Next
  </button>
</div>
`;

exports[`Snapshot tests renders nav buttons correctly for no buttons 1`] = `
<div
  className="NavButtonsstyles__PrevNext-tjoruy-0 fPtyJB"
>
  <button
    className="NavButtonsstyles__Previous-tjoruy-1 mipfr"
    disabled={true}
    onClick={[MockFunction]}
  >
    Previous
  </button>
  <button
    className="NavButtonsstyles__Next-tjoruy-2 eYzWCn"
    disabled={true}
    onClick={[MockFunction]}
  >
    Next
  </button>
</div>
`;

exports[`Snapshot tests renders results text correctly 1`] = `
<p
  className="Resultsstyles__ResultsCount-sc-1u314jh-0 komArg"
>
  Showing books 
  1
  -
  10
</p>
`;

exports[`Snapshot tests renders the app correctly 1`] = `
Array [
  <header>
    <h1
      className="Appstyles__Header-sc-1dagatm-0 jFTszl"
    >
      Book Search
    </h1>
  </header>,
  <main>
    <div
      className="Searchstyles__SearchDiv-sc-1afspv3-0 erVQWF"
    >
      <form
        className="Searchstyles__Form-sc-1afspv3-1 lmIVUu"
        onSubmit={[Function]}
      >
        <p
          className="Searchstyles__Title-sc-1afspv3-2 hXMlzD"
        >
          Search for a book by title or author
        </p>
        <select
          aria-label="Select title or author"
          className="Searchstyles__Select-sc-1afspv3-3 fOtYjG"
          data-testid="select"
          required={true}
        >
          <option
            className="hide"
            value=""
          >
            Title or author
          </option>
          <option
            value="intitle"
          >
            Title
          </option>
          <option
            value="inauthor"
          >
            Author
          </option>
        </select>
        <input
          aria-label="Type in the title or author you want to search for"
          className="Searchstyles__Input-sc-1afspv3-4 gbjkP"
          name="search"
          required={true}
          type="search"
        />
        <button
          className="Searchstyles__Submit-sc-1afspv3-5 kVqeMX"
        >
          Search
        </button>
      </form>
    </div>
    
    <div
      className="Appstyles__Books-sc-1dagatm-1 fkTpGD"
    />
  </main>,
]
`;

exports[`Snapshot tests renders the cards correctly 1`] = `
<div
  className="Cardstyles__CardDiv-sc-8m35hg-0 cLLdFS"
  data-testid="cardDiv"
>
  <img
    alt="Cover for Title 1"
    className="Cardstyles__Cover-sc-8m35hg-1 gvYeit"
    src="http://www.dummycover1.com"
  />
  <div
    className="Cardstyles__Info-sc-8m35hg-2 eIHPWj"
  >
    <p
      className="Cardstyles__Title-sc-8m35hg-3 bXzrjE"
      data-testid="title"
    >
      Title 1
    </p>
    <p
      className="Cardstyles__Author-sc-8m35hg-4 eHwNlE"
      data-testid="author"
    >
      by Author 1
    </p>
    <p
      className="Cardstyles__Publisher-sc-8m35hg-5 Vzfrb"
      data-testid="publisher"
    >
      Publisher: Publisher 1
    </p>
    <a
      href="https://www.dummylink1.com"
      title="More information about Title 1"
    >
      More information
    </a>
  </div>
</div>
`;

exports[`Snapshot tests renders the search correctly 1`] = `
<div
  className="Searchstyles__SearchDiv-sc-1afspv3-0 erVQWF"
>
  <form
    className="Searchstyles__Form-sc-1afspv3-1 lmIVUu"
    onSubmit={[Function]}
  >
    <p
      className="Searchstyles__Title-sc-1afspv3-2 hXMlzD"
    >
      Search for a book by title or author
    </p>
    <select
      aria-label="Select title or author"
      className="Searchstyles__Select-sc-1afspv3-3 fOtYjG"
      data-testid="select"
      required={true}
    >
      <option
        className="hide"
        value=""
      >
        Title or author
      </option>
      <option
        value="intitle"
      >
        Title
      </option>
      <option
        value="inauthor"
      >
        Author
      </option>
    </select>
    <input
      aria-label="Type in the title or author you want to search for"
      className="Searchstyles__Input-sc-1afspv3-4 gbjkP"
      name="search"
      required={true}
      type="search"
    />
    <button
      className="Searchstyles__Submit-sc-1afspv3-5 kVqeMX"
    >
      Search
    </button>
  </form>
</div>
`;
